eval.ml

```diff
@@ -199,34 +200,46 @@ let reduce conf =
        trace = v :: conf.trace;
        pc = pc';
      }
-  | Invalidate (e, l, xs) ->
+  | Osr (e, v, l, osr) ->
      let b = get_bool (eval conf e) in
-     if b then
+     if not b then
        { conf with
          pc = pc';
        }
      else begin
-       let add env x =
-         match Env.find x conf.env with
-         | exception Not_found -> raise (Unbound_variable x)
-         | v -> Env.add x v env
+       let add (env, heap) = function
+         | OsrConst (x, e) ->
+           (Env.add x (Const (eval conf e)) env,
+            heap)
+         | OsrMut (x, OsrExp e) ->
+           let a = Address.fresh () in
+           (Env.add x (Mut a) env,
+            Heap.add a (Value (eval conf e)) heap)
+         | OsrMut (x, OsrUndef) ->
+           let a = Address.fresh () in
+           (Env.add x (Mut a) env,
+            Heap.add a Undefined conf.heap)
```

Comment:
This is not an optimization, I believe it is important semantically -- it might be relevant for the proof as you also suggest. If two versions have *identical* instruction streams, I want to be able to insert an `osr` at some place that jumps to the same place in the other version, and this should be a no-op (besides the version change). This only works if you are able to say, like we merge two different constant declarations (morally, we have a phi-node), that we merge two different notions of mutable variable into one (and in particular, the heap cells are shared as well). Allocation on the osr also makes perfect sense semantically, but only in the case where the corresponding allocation was avoided in the incoming version in the first place, for example by sinking a mutable variable into a constant variable.